### PR TITLE
[FLINK-1967] Introduce (Event)time in Streaming

### PR DESCRIPTION
**Pull request by aljoscha**
## [FLINK-1967] Introduce (Event)time in Streaming
    
This introduces an additional timestamp field in StreamRecord. When using a
SourceFunction and an auto-timestamp interval is set using the
ExecutionConfig, the timestamp is automatically set to System.currentTimeMillis()
upon element emission. The timestamp can be manually set using an
EventTimeSourceFunction.

This also changes the signature of the StreamOperators. They now get
a StreamRecord instead of the unwrapped value. This is necessary for
them to access the timestamp. Before, the StreamTask would unwrap the
value from the StreamRecord, now this is moved one layer higher.

This also introduces watermarks to keep track of processing. At
a configurable interval the sources will emit watermarks that signify
that no records with a lower timestamp will be emitted in the future by
this source. The timestamps are broadcast, stream inputs wait for watermark
events on all input channels and forward the watermark to the
StreamOperator once the watermark advances on all inputs. Operators are
responsible for forwarding the watermark once they know that no elements
with a previous timestamp will be emitted (this is mostly relevant for
buffering operations such as windows). Right now, all operators simply
forward the watermark they receive.

When using an EventTimeSourceFunction the system does not
automatically emit timestamps, the user is required to manually emit
watermarks using the SourceContext.

No watermarks will be emitted unless
ExecutionConfig.setAutoWatermarkInterval is used to set a watermark
interval.

This commit contains fixes for other stuff that was discovered while
implementing the feature. See Jira issue numbers and descriptions below.

## [FLINK-2290/2295] Fix CoReader Event Handling and Checkpointing

This changes CoReader (now CoStreamingRecordReader) to reuse
UnionGate for the input multiplexing. This way it will not lock in on
one specific input side and read events from both input sides.

This also enables an event listener for checkpoint barriers so that the
TwoInputTask now reacts to those and correctly forwards them.

Then, this adds CoStreamCheckpointintITCase to verify that checkpointing
works in topologies with TwoInputStreamTasks.

This also adds tests for OneInputStreamTask and TwoInputStreamTask
that check whether:
 - whether open()/close() of RichFunctions are correctly
   called
 - Watermarks are correctly forwarded
 - Supersteps/checkpoint barriers are correctly forwarded and the
   blocking of inputs works correctly

## Add proper tests for Stream Operators

These test whether:
 - open()/close() on RichFunctions are called
 - Timestamps of emitted elements match the timestamp of the input
   element
 - Watermarks are correctly forwarded

## [FLINK-2301] Fix Behaviour of BarrierBuffer and add Tests

Before, a CheckpointBarrier from a more recent Checkpoint would also
trigger unblocking while waiting on an older CheckpointBarrier. Now,
a CheckpointBarrier from a newer checkpoint will unblock all channels
and start a new wait on the new Checkpoint.

The tests for OneInputStreamTask and TwoInputStreamTask check whether
the buffer behaves correctly when receiving CheckpointBarriers from more
recent checkpoints while still waiting on an older CheckpointBarrier.

### CODE REVIEW

**Comment by aljoscha** [comment]
This is a rework of #879. I reworked it to not use jodi-time anymore. Now timestamps are just long fields. Also, watermarks are now transmitted in band with the records, not as events.

I'm currently running some performance tests, once the results are in I'll post the numbers here.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/runtime/streamrecord/StreamRecordSerializer.java [hunk 1]
```diff
@@ -20,26 +20,35 @@
...
+ * {@link StreamRecord StreamRecords} and {@link Watermark Watermarks}. This serializer
+ * can handle both of them, therefore it returns {@link Object} the result has
+ * to be cast to the correct type.
+ *
+ * @param <T> The type of value in the {@link StreamRecord}
+ */
+public final class StreamRecordSerializer<T> extends TypeSerializer<Object> {
 
-public final class StreamRecordSerializer<T> extends TypeSerializer<StreamRecord<T>> {
+	private final long IS_WATERMARK = Long.MIN_VALUE;
```
Comment (gyfora) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/runtime/streamrecord/StreamRecordSerializer.java]: Why isn't this field a boolean? Just wondering :)

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+				if (result.isFullRecord()) {
+					Object recordOrWatermark = deserializationDelegate.getInstance();
+
+					if (recordOrWatermark instanceof Watermark) {
+						Watermark mark = (Watermark) recordOrWatermark;
+						long watermarkMillis = mark.getTimestamp();
+						if (watermarkMillis > watermarks[currentChannel]) {
+							watermarks[currentChannel] = watermarkMillis;
+							long newMinWatermark = Long.MAX_VALUE;
+							for (long watermark : watermarks) {
```
Comment (gyfora) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java]: I guess this wont work if we have an iteration, as watermarks will get stuck at the feedback point :/

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/runtime/streamrecord/StreamRecordSerializer.java (hunk 1 above)
Comment (aljoscha) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/runtime/streamrecord/StreamRecordSerializer.java]: This is used as a special timestamp. When the deserializer deserializes it it knows that the following long is a watermark. That's how the StreamRecordSerializer can be used for both watermarks and records. I'll add a comment.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java (hunk 2 above)
Comment (aljoscha) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java]: The whole thing with watermarks does not really work with iterations, correct. We'll have to figure out a way to do it.

**Comment by aljoscha** [comment]
These are the results of my performance tests: https://docs.google.com/spreadsheets/d/1mXcy25UOiZ4lEJOia1OM-K-U6np2bWv3x4q5nSnUMdg/edit?usp=sharing

Some numbers are maybe a bit inaccurate because I only gave it 6 minutes of runtime for every test configuration. Especially, with the watermark intervals, sometimes the throughput is better for a lower watermark interval. This seems counterintuitive but I think the performance can differ quite a bit between runs, I think @rmetzger has also seen this behavior.

**Comment by gyfora** [comment]
So one key question here is the overhead of the watermarks. (Throughput)

According to the spreadsheet this is more than 15% right?

**Comment by aljoscha** [comment]
The job I ran was source -> repartition -> sink, where the source is generating tuples in an infinite loop.

The gist of the data is this (I'm rerunning some of them with longer runtimes to remove outliers):
 - Introducing timestamp for each record that is set at source incurs about a 15 % performance hit in these tests 
 - Watermarks have no additional overhead once the interval is higher than 100 sec (I didn't test with lower intervals, such as 50 sec, but I will do these today.)

**Comment by aljoscha** [comment]
Oh and I forgot, the latency is not modified at all by these changes.

**Comment by aljoscha** [comment]
Would anyone like to look into the implementation? I think if we want to have this feature we should merge this rather sooner than later because it touches a lot of stuff and also contains fixes and lot of tests for other stuff.

**Comment by gyfora** [comment]
I think the implementation looks good however I am concerned about the performance hit.

We need to make sure that we dont get this serious overhead when the job does not even need timestamps. (No time windows)

**Comment by StephanEwen** [comment]
As far as I understand, the current mechanism ("operator time") is still possible. Then it would be okay.

Concerning always adding timestamps to stream records: That may be necessary anyways, if we want to keep track of latency - for monitoring.

**Comment by gyfora** [comment]
If we can still have the records without timestamps that would be good. But this should be automatic.

Regarding monitoring, this seems to be an overkill for that purpose. We could "sample" the stream and attach timestamps there and that wouldn't be a large overhead.

**Comment by aljoscha** [comment]
I could add a mode where the sources don't attach timestamps and where the StreamRecordSerializer ignores the timestamps and does not multiplex Records and Watermarks.

What do you mean by sampling the stream and attaching timestamps? If we want to attach timestamps then the whole machinery for dealing with them needs to be active anyways.

**Comment by gyfora** [comment]
I was just trying to make the point that for monitoring we don't need to attach timestamps for all the records. If monitoring means 15% less throughput I would probably not enable it anyways.

**Comment by StephanEwen** [comment]
Yeah, we can add a way to not have timestamps. But my feeling is that this is pre-mature optimization.

The point where this is really relevant is when people do super small records and push the throughput to max capacity. That impacts standalone benchmarks, but is probably not a real world issue.

**Comment by aljoscha** [comment]
I added numbers to my google sheet (https://docs.google.com/spreadsheets/d/1mXcy25UOiZ4lEJOia1OM-K-U6np2bWv3x4q5nSnUMdg/edit?usp=sharing) that show the performance with timestamps disabled in sources and the StreamRecordSerializer. Now I'm running tests with larger payloads to see how the impact of the timestamps diminished with growing element size.

**Comment by aljoscha** [comment]
I now added a switch that disables timestamps and watermark multiplexing. Timestamps and multiplexing are disabled by default, i.e. the performance of existing stuff is not changed.

**Comment by gyfora** [comment]
Thats great, thanks Aljoscha. Sorry for raising so many concerns. I just want to make sure this change will be smooth :)

**Comment by aljoscha** [comment]
No worries, that's why we have the process. :smiley_cat:

**Comment by aljoscha** [comment]
Does anyone have any objections still? The impact of the timestamps is now completely disabled by default.

If there are no objections I would like to merge this by tomorrow.

**Comment by gyfora** [comment]
I think you should go ahead with it. The sooner it is in the more testing it gets :)

**Comment by aljoscha** [comment]
Manually merged

**Pull request closed without merging**